aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -254,6 +254,157 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
   return at::linalg_matrix_rank(self, c10::nullopt, symmetric);
 }
 
+// multi_dot helper functions
+namespace {
+
+/**
+ * @brief Computes the optimal order to multiply the input tensors
+ *
+ * Follows the dynamic programming algorithm from Cormen et al,
+ * "Introduction to Algorithms, Third Edition", Chapter 15.2,
+ * p. 370-378. Note that the book uses 1-based indexing.
+ *
+ * @param tensors list of 2D tensors
+ * @return 2D vector used by #matrix_chain_multiplication
+ */
+std::vector<std::vector<int64_t>> matrix_chain_order(
+    const std::vector<Tensor>& tensors) {
+  const auto n = tensors.size();
+
+  // Tensor i has dimensions p[i] by p[i + 1]
+  std::vector<int64_t> p(n + 1);
+  for (auto i = decltype(n){0}; i < n; ++i) {
+    p[i] = tensors[i].size(0);
+  }
+  p[n] = tensors[n - 1].size(1);
+
+  // Memoization matrices for the cost and order
+  std::vector<std::vector<int64_t>> m(n, std::vector<int64_t>(n, 0));
+  std::vector<std::vector<int64_t>> s(n, std::vector<int64_t>(n));
+
+  // For each sublist [i, j] consider the cost of splitting the multiplication
+  // into tensors [0...i] and [j...n] and then multiplying the result of those.
+  for (auto l = decltype(n){1}; l < n; ++l) {
+    for (auto i = decltype(n){0}; i < n - l; ++i) {
+      const auto j = i + l;
+      m[i][j] = std::numeric_limits<int64_t>::max();
+      for (auto k = decltype(n){i}; k < j; ++k) {
+        const auto q = m[i][k] + m[k + 1][j] + p[i] * p[k + 1] * p[j + 1];
+        if (q < m[i][j]) {
+          m[i][j] = q;
+          s[i][j] = k;
+        }
+      }
+    }
+  }
+
+  return s;
+}
+
+// Multiplies the tensors following the order from #matrix_chain_order
+Tensor matrix_chain_multiplication(
+    const std::vector<Tensor>& tensors,
+    const std::vector<std::vector<int64_t>>& order,
+    int64_t i,
+    int64_t j) {
+  if (i == j) {
+    return tensors[i];
+  }
+  return at::mm(
+      matrix_chain_multiplication(tensors, order, i, order[i][j]),
```

Comment:
Can we write this iteratively?

Reply:
We can by maintaining explicit stacks but the algorithm is intrinsically recursive so this is the most natural expression of it in code. I think performance here is superseded by the multiplications.

Reply:
FYI `torch.chain_matmul` also uses recursion.

Reply:
Yes by maintaining explicit stacks but I don't think it's worth it as the perf bottleneck here is the mm calls.